File: advanced-programming-methods/toy-interpreter-java/src/model/SymTable.java
package model;

import exceptions.DictionaryException;

/**
 * The type Sym table.
 */
public class SymTable implements IDictionary
{
    private Pair[] pairs;
    private int size = 0;

    /**
     *
     * @param key
     * @return
     * @throws DictionaryException
     */
    private Pair findPair(String key) throws DictionaryException
    {
        for (int i = 0; i < size; i++)
        {
            Pair pair = pairs[i];
            if (pair.getKey().equals(key))
            {
                return pair;
            }
        }

        throw new DictionaryException("Pair not found");
    }

    /** Constructs a new SymTable. */
    public SymTable()
    {
        pairs = new Pair[Constants.MAX_SIZE];
    }

    /** {@inheritDoc} */
    public int get(String key) throws DictionaryException
    {
        Pair pair;

        try
        {
            pair = findPair(key);
        }
        catch (DictionaryException e)
        {
            throw new DictionaryException(String.format("Key = '%s' does not exist", key));
        }

        return pair.getValue();
    }

    public String getAllSave()
    {
        return "NOT IMPLEMENTED\n";
    }

    /** {@inheritDoc} */
    public void put(String key, int value)
    {
        Pair foundPair = null;

        try
        {
            foundPair = findPair(key);
        }
        catch (DictionaryException e)
        {
        }

        if (foundPair != null)
        {
            foundPair.setValue(value);
        }
        else
        {
            // key does not already exist
            Pair pair = new Pair(key, value);

            pairs[size++] = pair;
        }
    }

    @Override
    public IDictionary clone()
    {
        IDictionary dictionary = new SymTable();

        for (int i = 0; i < size; i++)
        {
            Pair pair = pairs[i];
            dictionary.put(pair.getKey(), pair.getValue());
        }

        return dictionary;
    }

    /** {@inheritDoc} */
    public void clear()
    {
        pairs = new Pair[Constants.MAX_SIZE];
        size = 0;
    }

    /** {@inheritDoc} */
    @Override
    public String toString()
    {
        if (size == 0)
        {
            return "{}";
        }

        String out = "{";

        for (int i = 0; i < size - 1; i++)
        {
            out += pairs[i].toString() + ", ";
        }
        out += pairs[size - 1].toString() + "}";

        return out;
    }
}

File: advanced-programming-methods/labs/toy-interpreter-sharp/model/SymTable.cs
using System;
using Exceptions;

namespace Model
{
    /// <summary>
    /// The type Sym table.
    /// </summary>
    public class SymTable : IDictionary
    {
        private Pair[] pairs;
        private int size = 0;

        /// 
        /// <param name="key">
        /// @return </param>
        /// <exception cref="DictionaryException"> </exception>
        private Pair FindPair(string key)
        {
            for (int i = 0; i < size; i++)
            {
                Pair pair = pairs[i];
                if (pair.Key.Equals(key))
                {
                    return pair;
                }
            }

            throw new DictionaryException("Pair not found");
        }

        public string AllSave
        {
            get
            {
                return "";
            }
        }

        /// <summary> Constructs a new SymTable. </summary>
        public SymTable()
        {
            pairs = new Pair[Constants.MAX_SIZE];
        }

        /// <inheritdoc />
        public int Get(string key)
        {
            Pair pair;

            try
            {
                pair = FindPair(key);
            }
            catch (DictionaryException)
            {
                throw new DictionaryException(string.Format("Key = '{0}' does not exist", key));
            }

            return pair.Value;
        }

        /// <inheritdoc />
        public void Put(string key, int value)
        {
            Pair pair = null;

            try
            {
                pair = FindPair(key);
            }
            catch (DictionaryException)
            {

            }

            if (pair != null)
            {
                pair.Value = value;
            }
            else
            {
                Pair pairi = new Pair(key, value);

                pairs[size++] = pairi;
            }
        }

        public IDictionary Clone()
        {
            IDictionary dictionary = new SymTable();

            for (int i = 0; i < size; i++)
            {
                Pair pair = pairs[i];
                dictionary.Put(pair.Key, pair.Value);
            }

            return dictionary;
        }

        /// <inheritdoc />
        public void Clear()
        {
            pairs = new Pair[Constants.MAX_SIZE];
            size = 0;
        }
            
        public override string ToString()
        {
            if (size == 0)
            {
                return "{}";
            }

            string @out = "{";

            for (int i = 0; i < size - 1; i++)
            {
                @out += pairs[i] + ", ";
            }
            @out += pairs[size - 1] + "}";

            return @out;
        }
    }
}

